### ISSUE TITLE
Dealing with thousands of errors (Or, event_from_exception is slow. Or, Rails 7's error_reporter is awkward)

### ISSUE THREAD

**Issue by jdelStrother** [getsentry/sentry-ruby#1765]
We recently had a cache server go down, which caused our servers to become overwhelmed and start dropping requests like crazy. This seemed unusual: previously we've been fine with a cache server disappearing - we might get a handful of errors, but Rails' cache normally behaves fairly sensibly, just behaving like a cache-miss & silently failing on writes.

I think the new problem lies in a mix of Rails 7's error reporting behaviour, and Sentry's event_from_exception being a relatively slow method.

In Rails 7, every single cache error gets sent to the error_reporter (https://github.com/rails/rails/blob/0169d15bc7ec4557971d6ac6120e48b2cac9c407/activesupport/lib/active_support/cache/redis_cache_store.rb#L460-L466)

These then get sent to Sentry. We're using Sentry's BackgroundWorker to process these asynchronously (and drop events that exceed the `max_queue` size), but before that happens, Sentry calls event_from_exception on the current thread. This seems like quite a heavy method, mostly due to StackTraceBuilder - 

<img width="1452" alt="image" src="https://user-images.githubusercontent.com/2377/158985235-26455533-f177-4b97-ac4a-5cf3db5255bc.png">

and worse, it's called for every single error-report, regardless if BackgroundWorker is about to discard the error because the queue is full. `config.before_send` is called after event_from_exception, so there's not an obvious way for me to control whether that work gets done.

I'm not sure what a good fix would look like here - 
* Possibly the fault lies with Rails 7 - for now, I think I'm going to patch the RedisCacheStore#failsafe method to not call `error_reporter&.report`.  It does seem unfortunate that even if you provide your own `error_handler` callback, ActiveSupport.error_reporter gets called regardless.
* It would be nice if `event_from_exception` happened asynchronously in BackgroundWorker - that way a) it's not blocking the current thread, and b) it would benefit from not doing unnecessary work when we're just about to discard the exception.  But that looks like quite a big change from the current flow, especially if it's keeping compatibility with the other `config.async` options.
* Maybe Sentry.capture_exception could have its own throttle that just discards exceptions if there's been more than X in the past Y seconds ?

Any other suggestions?

**Comment by st0012** [comment]
Thanks for reporting this, but I have a few questions before I can provide suggestions:

1. From the chart, it looks like `Client#event_from_exception` takes 4.5 secs? Are you able to create this slowness locally? Or do you think it's a result of resource contention on production?
    - I think one possible scenario is that one exception has deeply-nested causes (prob 20+ in total). That may make Event initialization very slow. But this isn't what we expect to have in normal applications.
2. Are the captured Redis exceptions helpful to you?

**Comment by sl0thentr0py** [comment]
@jdelStrother thx for the great issue!

There are several things going on here.
1. A cache failure getting reported as an 'error' somewhat breaks our 'error' semantic assumptions which are only generally  exceptions. This is more an implicit understanding of what an 'error' is which is of course open to some debate.
2. There is a general scale assumption issue. If we have some error which happens at a rate much faster than the rate at which we can process the `BackgroundWorker` transport sending to Sentry, we will always run into such problems. This is something we should certainly think about and I will keep this in mind.
3. As @st0012 just said, `event_from_exception` taking that long is sort of suspicious and seems the most time is spent in `longest_load_path`. So you could either have a very long `$LOAD_PATH` somehow or the string `starts_with?` and `max` logic is taking too long. We could look into optimizing the algo here but not sure if it's really the problem.
4. `event_from_exception` always running even if the error will be discarded either due to a full queue or sample rate etc. We could again keep this in mind for a future refactor so that as minimum work is done as possible but I still think the full queue problem is hard to solve because we want to do that logic on the transport level.

**Comment by jdelStrother** [comment]
> From the chart, it looks like Client#event_from_exception takes 4.5 secs? Are you able to create this slowness locally? Or do you think it's a result of resource contention on production?

Yeah, sorry - this was an artificial benchmark wrapped in `1000.times{}`, partly because I was trying to measure the time taken once the BackgroundWorker queue was full.  Actual time is about 4 - 5ms.
(Which isn't _very_ slow, but gets bad if you're calling Rails.cache.read multiple times in a request and every single one hits the error reporter)

> I think one possible scenario is that one exception has deeply-nested causes (prob 20+ in total). That may make Event initialization very slow. But this isn't what we expect to have in normal applications.

Looks like I only have a single `cause` - the exception is Redis::CannotConnectError with a cause of Errno::ECONNREFUSED.

> Are the captured Redis exceptions helpful to you?

It'd be useful to be alerted if our cache is down, but it's not useful to have the hundreds of errors per second that this incident was generating.  I'm not sure there's much useful context in any given exception either - eg the stacktrace or captured local variables probably aren't going to help me.

> As @st0012 just said, event_from_exception taking that long is sort of suspicious and seems the most time is spent in longest_load_path. So you could either have a very long $LOAD_PATH somehow or the string starts_with? and max logic is taking too long. We could look into optimizing the algo here but not sure if it's really the problem.

LOAD_PATH has 271 items, the Redis::CannotConnectError has 62 frames , and another 63 frames on the `cause`.  And yeah, longest_load_path looks like the culprit: 

```
irb(main):053:0> Benchmark.ms { (ex.backtrace + ex.cause.backtrace).each {|abs_path|
  $LOAD_PATH.select { |path| abs_path.start_with?(path.to_s) }.max_by(&:size)
} }
=> 4.833262006286532
```

I can share the actual LOAD_PATH & error backtraces privately if it'll help, though you might also be able to reproduce locally with something like:


```ruby
cache = ActiveSupport::Cache::RedisCacheStore.new(url: 'redis://0.0.0.0:1')
cache.read('x')
```

**Comment by st0012** [comment]
@jdelStrother can you do another benchmark that involves an actual request? and 4~5 ms seems reasonable to me. imo, if a rack request normally take 50ms to complete now takes 65ms (triggered 3 redis error), it's still acceptable?

I think the main problem is that the redis exception is reported a lot more frequently than we'd want it to be. but the SDK actually doesn't have control on that.

so as I explained in https://github.com/rails/rails/pull/43625#issuecomment-1072514175, I'll make the integration opt-in only in the next minor release. that way you won't need to deal with this new type of error that comes with a large volume.

**Comment by jdelStrother** [comment]
> I think the main problem is that the redis exception is reported a lot more frequently than we'd want it to be. but the SDK actually doesn't have control on that.

Totally agree that this is made 100x worse by Rails7's behaviour here, and we're going to monkeypatch `RedisCacheStore#failsafe` to deal with it.

Still, in real requests, we had some high-traffic requests that normally take 20ms jump to 120ms. They were triggering the error about 6-8 times each, mostly from Rack::Attack checks.
Another awful case is our RSS feeds that might have 400 partial-renders, each of which triggers a cache-read (and so 400x capture_exceptions).


Still, weird Rails behaviour aside, it would be very unusual for us to report more than 1 exception to Sentry per request - we'd typically only report an error if we're about to render a 500 response and then bail.

**Comment by st0012** [comment]
can you tell me more about the Rack::Attack case? I thought it'd only raise on exception, get reported, and then return the response? why would it have 6~8 exceptions?

**Comment by st0012** [comment]
@jdelStrother ah I see. so that'll also go away if the redis error is not reported anymore.
